torch/fx/experimental/fuser.py

```diff
@@ -62,3 +70,279 @@ def fuse(model: torch.nn.Module, inplace=False) -> torch.nn.Module:
+    model: torch.nn.Module,
+    use_mkl_heuristic: Callable[[MklSubgraph], bool] = use_mkl_length
+) -> torch.nn.Module:
+    """
+    Performs a set of optimization passes to optimize a model for the
+    purposes of inference. Specifically, the passes that are run are:
+    1. Conv/BN fusion
+    2. Dropout removal
+    3. MKL layout optimizations
+
+    The third optimization takes a function `use_mkl_heuristic` that's used
+    to determine whether a subgraph should be explicity run in MKL layout.
+    """
+    model = fuse(model)
+    model = remove_dropout(model)
+    fx_model = fx.symbolic_trace(copy.deepcopy(model))
+    modules: Dict[str, nn.Module] = dict(fx_model.named_modules())
+
+    # Inserts to_mkldnn and to_dense around every node we want to be a MKLDNN node.
+    # If the op is in `mkldnn_supported` then we always treat it as a MKLDNN node.
+    # However, if it's in `mkldnn_supported_unknown`, then we only treat it as
+    # a MKLDNN node if its inputs are MKLDNN nodes.
+    for node in list(fx_model.graph.nodes):
+        supports_mkldnn = 'no'
+        if node.op == 'call_module':
+            cur_module = modules[node.target]
+            if type(cur_module) in mkldnn_supported:
+                supports_mkldnn = 'yes'
+        elif node.op == 'call_function':
+            if node.target in mkldnn_supported:
+                supports_mkldnn = 'yes'
+            elif node.target in mkldnn_supported_unknown:
+                supports_mkldnn = 'unknown'
+
+        if supports_mkldnn != 'no':
+            if supports_mkldnn == 'unknown':
+                if not any([arg.target == 'to_dense' for arg in node.args]):
+                    continue
+            with fx_model.graph.inserting_before(node):
+                mkldnn_args = fx.map_arg(node.args, lambda n: fx_model.graph.call_method('to_mkldnn', (n, )))
+
+            node.args = cast(Tuple[fx.node.Argument], mkldnn_args)
+
+            with fx_model.graph.inserting_after(node):
+                dense_x = fx_model.graph.create_node('call_method', 'to_dense', (node,))
+                node.replace_all_uses_with(dense_x)
+                dense_x.args = (node,)
+
+    # Does pre-conversion of all modules into MKLDNN (when possible)
+    modules_to_mkldnn(list(fx_model.graph.nodes), modules)
+
+    # optimizes all a -> to_dense -> to_mkldnn -> b patterns into a -> b
+    for node in fx_model.graph.nodes:
+        if node.op == 'call_method' and node.target == 'to_dense':
+            prv_node = node.args[0]
+            users = list(node.users)
+            for user in users:
+                if user.op == 'call_method' and user.target == 'to_mkldnn':
+                    user.replace_all_uses_with(prv_node)
+                    fx_model.graph.erase_node(user)
+            if len(node.users) == 0:
+                fx_model.graph.erase_node(node)
+
+    fx_model.recompile()
+
+    num_nodes = len(fx_model.graph.nodes)
+    uf = UnionFind(num_nodes)
+
+    def get_color(n):
```

Comment:
Add a comment defining what each of `color`, `start_color` and `end_color` mean when found on a node